src/main/scala/firrtl/analyses/IRLookup.scala

```diff
@@ -0,0 +1,343 @@
+// See LICENSE for license details.
+
+package firrtl.analyses
+
+import firrtl.annotations.TargetToken._
+import firrtl.annotations._
+import firrtl.ir._
+import firrtl.passes.MemPortUtils
+import firrtl.{DuplexFlow, ExpKind, SinkFlow, Flow, InstanceKind, Kind, SourceFlow, MemKind, PortKind, RegKind, UnknownFlow, Utils, WInvalid, WireKind}
+
+import scala.collection.mutable
+
+object IRLookup {
+  def apply(circuit: Circuit): IRLookup = ConnectionGraph(circuit).irLookup
+}
+
+/** Handy lookup for obtaining AST information about a given Target
+  * @param declarations Maps references (not subreferences) to declarations
+  * @param modules Maps module targets to modules
+  */
+class IRLookup private[analyses] ( private val declarations: mutable.LinkedHashMap[ModuleTarget, mutable.LinkedHashMap[ReferenceTarget, FirrtlNode]],
```

Comment:
I believe @jackkoenig ran into a performance issue where the hashmaps got too big. We should just keep an eye out for that, especially for the `flowCache`, `kindCache`, `tpeCache` and `exprCache`. We may want to guard the lookups with the `ModuleTarget` first.

Reply:
In my next PRs I think I should rewrite IRLookup construction.

Reply:
Currently, `IRLookup` is constructed by BFS when constructing `ConnectionGraph`, I think:
1. we can use simple linear search to visit AST to construct `IRLookup`, parallel level will be much higher, but need additional search.
2. when visiting AST, we can directly generate flow, kind, tpe, expr etc Map, rather using cache to consult, but it will lead to huge memory usage. which is your prefer? the mutable maps in IRLookup are so scary.
3. performance issue you mentioned has been fixed by ModuleTarget guarding